lib/internal/process/promises.js

```diff
@@ -133,18 +153,62 @@ function processPromiseRejections() {
   while (len--) {
     const promise = pendingUnhandledRejections.shift();
     const promiseInfo = maybeUnhandledPromises.get(promise);
-    if (promiseInfo !== undefined) {
-      promiseInfo.warned = true;
-      const { reason, uid } = promiseInfo;
-      if (!process.emit('unhandledRejection', reason, promise)) {
-        emitWarning(uid, reason);
-      }
-      maybeScheduledTicks = true;
+    if (promiseInfo === undefined) {
+      continue;
+    }
+    promiseInfo.warned = true;
+    const { reason, uid } = promiseInfo;
+    if (state === states.strict) {
+      fatalException(reason);
+    }
+    if (!process.emit('unhandledRejection', reason, promise)) {
+      emitWarning(uid, reason);
     }
+    maybeScheduledTicks = true;
   }
   return maybeScheduledTicks || pendingUnhandledRejections.length !== 0;
 }
 
+function getError(name, message) {
+  // Reset the stack to prevent any overhead.
+  const tmp = Error.stackTraceLimit;
+  Error.stackTraceLimit = 0;
+  // eslint-disable-next-line no-restricted-syntax
+  const err = new Error(message);
+  Error.stackTraceLimit = tmp;
+  Object.defineProperty(err, 'name', {
+    value: name,
+    enumerable: false,
+    writable: true,
+    configurable: true,
+  });
+  return err;
+}
+
+function fatalException(reason) {
+  let err;
+  if (reason instanceof Error) {
+    err = reason;
+  } else {
+    err = getError(
+      'UnhandledPromiseRejection',
+      'This error originated either by ' +
+        'throwing inside of an async function without a catch block, ' +
+        'or by rejecting a promise which was not handled with .catch().' +
+        ` The promise rejected with the reason "${safeToString(reason)}".`
+    );
+    err.code = 'ERR_UNHANDLED_REJECTION';
+  }
+  // Throw an error if the user did not handle the exception.
+  if (!process._fatalException(err, 'unhandledRejection')) {
```

Comment:
we should probably document this, but _fatalException is actually a child of the real fatal exception handling that happens in node.cc. you should call `internalBinding('util').triggerFatalException` here.

Reply:
Calling `triggerFatalException` also matches the behavior of unhandled exceptions on a microtask, and keeping the behavior similar (in strict mode) is a good idea IMO.

Reply:
Although the stack trace from `triggerFatalException` with `--abort-on-uncaught-exception` is rather useless 😕

![image](https://user-images.githubusercontent.com/4048656/54159825-b35d5980-442c-11e9-9e19-cb88dcbfb203.png)

But this can be improved in a separate, unrelated pull request.

Reply:
While switching to `triggerFatalException` I also fixed printing the exception on abort.

![image](https://user-images.githubusercontent.com/8822573/54202573-a2feba80-44d0-11e9-9116-a9756dd49b07.png)

Reply:
@devsnek PTAL as I marked the fatal exception triggered in esm as promise.